[DhiraPT] iP

DhiraPT opened the pull request:
# Eggy
> “Better an egg today than a hen tomorrow.” 
~ François Rabelais, Gargantua and Pantagruel ([source](https://en.wiktionary.org/wiki/better_an_egg_today_than_a_hen_tomorrow))

### Eggy is
- Cute
- Can remember all your tasks
- ~~PRETTY~~ _**SUPER**_ lightweight

### Follow these steps to start chatting with Eggy:
1. Download it from [here](https://github.com/DhiraPT/ip/releases)
2. Follow the steps to run
3. Start adding tasks
4. Lastly, have fun with Eggy! :partying_face:

### Features:
- [x] Managing tasks (todo, deadline, event)
- [x] Add reminders (coming soon)
- [ ] AI-powered (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. 
Eggy's life starts from this method in the `Launcher` class:
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

DhiraPT committed:
Rename bot and add greeting and goodbye

DhiraPT committed:
Add user inputs

DhiraPT committed:
Allow add task and return task list

DhiraPT committed:
Add Task class and implement mark and unmark task

DhiraPT committed:
Add 3 types of tasks

DhiraPT committed:
Add automated text UI testing

DhiraPT committed:
Add exceptions to handle errors and set Task class to abstract

DhiraPT committed:
Update automated test for exceptions

DhiraPT committed:
Add delete task, add IncompleteCommandException, and clean the code

DhiraPT committed:
Use enums for commandType and add TaskListIndexOutOfBoundsException

DhiraPT committed:
TaskNumberFormatException

DhiraPT committed:
Implement save to hard disk functionality

DhiraPT committed:
Merge branch 'branch-Level-7'

DhiraPT committed:
Add date and time parsing

DhiraPT committed:
Merge branch 'branch-Level-8'

DhiraPT committed:
Modify the code into OOP

DhiraPT committed:
Modify the code into OOP

DhiraPT committed:
Clean up code and fix errors

DhiraPT committed:
Merge branch 'branch-A-MoreOOP'

DhiraPT committed:
Organize into packages

DhiraPT committed:
Merge branch 'branch-A-Packages'

DhiraPT committed:
Merge remote-tracking branch 'origin/add-gradle-support'

DhiraPT committed:
Modify build.gradle

DhiraPT committed:
Add JUnit Tests

DhiraPT committed:
Merge branch 'branch-A-JUnit'

DhiraPT committed:
Add DateTimeFormatException

DhiraPT committed:
Add configuration for build file

DhiraPT committed:
Add JavaDoc

DhiraPT committed:
Soft wrap the lines

DhiraPT committed:
Implement find macthing task

DhiraPT committed:
Implement find macthing task

DhiraPT committed:
Merge branch 'branch-A-JavaDoc'

DhiraPT committed:
Merge branch 'branch-A-CodingStandard'

DhiraPT committed:
Soft wrap lines and fix spelling

DhiraPT committed:
Resolve conflicts with branch-Level-9

sdevih commented on src/main/java/eggy/task/Task.java:
```diff
@@ -44,6 +44,15 @@ public void unmarkDone() {
+     *
+     * @return The name of the task.
+     */
+    public String getName() {
+        return this.name;
+    }
+
```
I like that you adhered to the naming conventions of methods by using verbs and camelCase!

sdevih commented on src/main/java/eggy/command/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+package eggy.command;
+
+import eggy.exception.IncompleteTaskException;
+import eggy.storage.Storage;
+import eggy.task.TaskList;
+import eggy.task.Task;
+import eggy.ui.Ui;
```
I like that imports are explicitly named and clearly show what's used in the class!

sdevih commented on src/main/java/eggy/ui/Ui.java:
```diff
@@ -109,6 +110,18 @@ public void printTaskListEmpty() {
+     *
+     * @param tasks The matching tasks.
+     */
+    public void printMatchingTasks(List<Task> tasks) {
+        System.out.println("     Here are the matching tasks in your list:");
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println("     " + (i + 1) + "." + tasks.get(i));
+        }
+    }
```
May I ask why there's a space before printing the corresponding task number and description?

DhiraPT replied:
I followed the exact indentation from the project guidelines example.

sdevih commented on src/main/java/eggy/task/TaskList.java:
```diff
@@ -73,6 +74,16 @@ public boolean isEmpty() {
+     * @param keyword The keyword to search for.
+     * @return List of tasks that match the keyword.
+     */
+    public List<Task> findMatchingTasks(String keyword) {
+        return this.tasks.stream().filter(task -> task.getName().contains(keyword)).collect(Collectors.toList());
+    }
+
```
I liked how you distinguished between single-valued and multi-valued variables, by using the plural form for the method name since it returns a list of tasks.

sdevih reviewed (commented):
LGTM! Overall a great job done with little to no nits to fixes. Keep up the good work!

Vision-2000 commented on src/test/java/eggy/storage/StorageTest.java:
```diff
@@ -0,0 +1,77 @@
+import eggy.task.Event;
+import eggy.task.Task;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.util.List;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
I like the clear organisations of imports by packages and alphabetical order, making it easier to remember.

Vision-2000 commented on src/main/java/eggy/ui/Ui.java:
```diff
@@ -0,0 +1,133 @@
+     * @param task The task removed.
+     * @param taskCount The number of tasks in the task list.
+     */
+    public void printTaskRemoved(Task task, int taskCount) {
+        System.out.println("     Noted. I've removed this task:");
+        System.out.println("       " + task.toString());
+        System.out.println("     Now you have " + taskCount + " tasks in the list.");
+    }
```
I like the naming of methods in this class, their objective is clear from the names alone without having to look at the code in detail.

Vision-2000 commented on src/main/java/eggy/parser/Parser.java:
```diff
@@ -0,0 +1,63 @@
+        String[] commands = fullCommand.split(" ", 2);
+        try {
+            CommandType commandType = CommandType.valueOf(commands[0].toUpperCase());
+            switch (commandType) {
+            case BYE:
+                return new ExitCommand();
+            case LIST:
+                return new ListCommand();
+            case DELETE:
+                return new DeleteCommand(commands, tasksSize);
+            case MARK:
+                return new MarkCommand(commands, tasksSize);
```
I like the use of `enum` in conjunction with `switch` statement, with the `case` clauses formatted correctly.

Vision-2000 commented on src/main/java/eggy/Eggy.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class Eggy {
+    /** The name of the chatbot. */
+    public final String NAME = "Eggy";
```
Should a static modifier be added since `NAME` indicates that it is a constant?

DhiraPT replied:
I think there might be some confusion between `final` and `static`. `static` doesn't imply constant.

DhiraPT replied:
Yes, i think `static` can be added here since `NAME` is not associated with instances of the class.

Vision-2000 reviewed (commented):
LGTM. Really well written code.

DhiraPT committed:
Implement checkstyle

DhiraPT committed:
Add Exception for FindCommand

DhiraPT committed:
Merge branch 'branch-A-CheckStyle'

DhiraPT committed:
Add JavaFX and fix StorageTest

DhiraPT committed:
Implement JavaFX GUI

DhiraPT committed:
Refractor GUI

DhiraPT committed:
Merge branch 'branch-Level-10'

DhiraPT committed:
Use varargs

DhiraPT committed:
Merge branch 'branch-A-Varargs'

DhiraPT committed:
Update the build version to v0.2
The build version is v0.1.

The user interface has been upgraded from text-based to a JavaFX-based GUI and is now ready for a new release.

Let's modify the build version to v0.2 in build.gradle.

This change will enable Eggy.jar to be built with the filename reflecting the latest version.

DhiraPT committed:
Add Java assertions to Storage class
The Storage class has no explicit checks to ensure critical conditions are met during saving or loading tasks.

This absence of checks may lead to errors during these operations.

Let's integrate assert statements into the load() and save() methods of the Storage class to validate important conditions during task loading and saving operations.

Adding Java assertions ensures that critical conditions, such as the validity of task data during loading and the non-nullness of the task list during saving, are enforced. This helps catch potential bugs early during development and testing phases.

DhiraPT committed:
Add Java assertions to MainWindow and DialogBox classes

DhiraPT committed:
Refractor code to avoid nesting and improve readability

DhiraPT committed:
Refractor code to further avoid nesting and code duplicates

DhiraPT committed:
Use Java Streams in the TaskList class
The toString() and the toFileString() method in the TaskList class are constructed using traditional for loop iteration, appending each task's formatted string to build the returned string.

This approach is verbose and less expressive.

Let's refactor these methods using Stream API to transform the list of tasks into a formatted string representation.

Stream API is chosen for its simplicity and readability in transforming collections of data. It also aligns with modern Java coding practices, enhancing code maintainability and readability.

DhiraPT committed:
Use Java Streams in the Response class

DhiraPT committed:
Merge pull request #1 from DhiraPT/branch-A-Assertions

Add Java assertions

DhiraPT committed:
Merge pull request #3 from DhiraPT/branch-A-Streams

Use Java Streams

DhiraPT committed:
Resolve conflicts with master branch

DhiraPT committed:
Merge pull request #2 from DhiraPT/branch-A-CodeQuality

Improve Code Quality

DhiraPT committed:
Add GitHub Actions

DhiraPT committed:
Deal with duplicate tasks

DhiraPT committed:
Merge branch 'branch-C-DetectDuplicates'

DhiraPT committed:
Fix dialogBox vertical truncation

The text within dialogBox gets cut when it extends beyong the visible
area, potentially causing users to miss important parts of the message.

Let's change the dialogContainer to use minHeight instead of prefHeight.

Using minHeight in dialogContainer will allow its children (dialogBox)
to dynamically adjust their height based on the length of the text
content, ensuring that all messages are fully visible.

DhiraPT committed:
Implement different colors for Eggy's dialogBox

DhiraPT committed:
Merge branch 'branch-A-BetterGui'

DhiraPT committed:
Change User picture

DhiraPT committed:
Change User picture

DhiraPT committed:
Update README.md and upload screenshot

DhiraPT committed:
Update README.md

DhiraPT committed:
Update build version to v1.0
